packages/astro/src/assets/fonts/utils.ts

```diff
@@ -264,3 +278,92 @@ export function kebab(value: string) {
 export function getFamilyName(family: Pick<FontFamilyAttributes, 'name' | 'as'>): string {
 	return family.as ?? family.name;
 }
+
+/**
+ * Resolves the font family provider. If none is provided, it will infer the provider as
+ * one of the built-in providers and resolve it. The most important part is that if a
+ * provider is not provided but `src` is, then it's inferred as the local provider.
+ */
+export async function resolveFontFamily({
+	family,
+	...resolveProviderOptions
+}: Omit<ResolveProviderOptions, 'provider'> & {
+	family: FontFamily<BuiltInProvider | FontProvider>;
+}): Promise<ResolvedFontFamily> {
+	if (family.provider === LOCAL_PROVIDER_NAME || (!family.provider && 'src' in family)) {
+		return {
+			...family,
+			provider: LOCAL_PROVIDER_NAME,
+		};
+	}
+
+	const provider =
+		family.provider === GOOGLE_PROVIDER_NAME || !family.provider ? google() : family.provider;
+
+	return {
+		...family,
+		provider: await resolveProvider({
+			...resolveProviderOptions,
+			provider,
+		}),
+	};
+}
+
+function sortObjectByKey<T extends Record<string, any>>(unordered: T): T {
+	const ordered = Object.keys(unordered)
+		.sort()
+		.reduce((obj, key) => {
+			// @ts-expect-error Type 'T' is generic and can only be indexed for reading. That's fine here
+			obj[key] = unordered[key];
+			return obj;
+		}, {} as T);
+	return ordered;
+}
+
+/**
+ * Extracts providers from families so they can be consumed by unifont.
+ * It deduplicates them based on their config and provider name:
+ * - If several families use the same provider (by value, not by reference), we only use one provider
+ * - If one provider is used with different settings for 2 families, we make sure there are kept as 2 providers
+ */
+export function familiesToUnifontProviders({
+	families,
+	hashString,
+}: {
+	families: Array<ResolvedFontFamily>;
+	hashString: (value: string) => string;
+}): Array<unifont.Provider> {
+	const hashes = new Set<string>();
+	const providers: Array<unifont.Provider> = [];
+
+	for (const { provider } of families) {
+		if (provider === LOCAL_PROVIDER_NAME) {
+			continue;
+		}
+
+		const unifontProvider = provider.provider(provider.config);
+		const hash = hashString(
+			JSON.stringify(
+				sortObjectByKey({
+					name: unifontProvider._name,
+					...provider.config,
+				}),
+			),
+		);
+		if (hashes.has(hash)) {
+			continue;
+		}
+		// Makes sure every font uses the right instance of a given provider
+		// if this provider is provided several times with different options
+		// We have to mutate the unifont provider name because unifont deduplicates
+		// based on the name.
+		unifontProvider._name += `-${hash}`;
+		// We set the provider name so we can tell unifont what provider to use when
+		// resolving font faces
+		provider.name = unifontProvider._name;
```

Comment:
I still don't know if this does something. This comes from `families` in the form of an object by the caller. Wouldn't it be a copy? 

If it does something, would it be possible to make the method pure and avoid side effects?

Reply:
I leave that to you. Please consider the following aspects:
While they both achieve the same result, the following approach **is a side effect** and *sometimes* difficult to test and/or predict.
- A side effect should be documented and/or expected by a maintainer
- Generally, throughout the code base, we tend to use a more pure/functional approach

Reply:
Updated